### PR TITLE
Aimed shot and spotting now have laser beams.

### PR DESCRIPTION
**Pull request by carlarctg**
# About the pull request

Aiming a shot on a sniper rifle now adds a red laser beam that connects the target and the aimer. Spotting designators will emanate a faint yellow one. Both will slowly increase in alpha from 0 to max as the tracking progresses.

The sniper's laser beam can be disabled with a button, but keeping it enabled reduces aiming shot time. Cloaked spotters will not cause a beam to appear.

Completely resprited the indicators for spotting (it's now yellow) and aiming (it's now red, usually)

The XM42B and PMC sniper rifles have more intense blue lasers and reticles.

Stepping through laser beams will check for eye protection, if it isn't sufficient, and the small probability chance passes, the target will scream in pain! Completely flavor. If they have enough eye protection, it will bounce off their headgear, and if said protection is BiMex patented personal shades it will create a cool visual effect.

Refactored eye_protection to have three levels, flavor, flashes, and welding. Sunglasses have flavor protection, sechuds have flash protection, welding gear has welding protection.

The spotter's designator now has a white band on it to distinguish it from the rest.

# Explain why it's good for the game

> Aiming a shot on a sniper rifle now adds a red laser beam that connects the target and the aimer. Spotting designators will emanate a faint yellow one. Both will slowly increase in alpha from 0 to max as the tracking progresses.

The primary reason for this addition is because it looks sick as FUCK. The secondary one is to help xenos see where they're being shot from.

> The sniper's laser beam can be disabled with a button, but keeping it enabled reduces aiming shot time. Cloaked spotters will not cause a beam to appear.

Makes sense, cloaked sniper/spotters shouldn't be made completely useless due to the beam. It makes sense for the laser to reduce aim time and it isn't a pain for xenomorphs to deal with as they know exactly where it is coming from and where they can run for cover (And other xenos can protect them better)

> Completely resprited the indicators for spotting (it's now yellow) and aiming (it's now red, usually)

It made no sense for sniper reticles to be blue. Red is danger, yellow is warning, blue is.... good? Additionally, since both reticles for spotting and aiming were nigh-identical, it caused quite the amount of confusion. I've tested it myself, you can spook away xenos with only spotting because they think it's the aimed shot reticle.

> The XM42B and PMC sniper rifles have more intense blue lasers and reticles.

Look, I know what I just said makes this sound stupid. But I think it makes a lot of sense for the dangerous high-tech sniper rifles to have stronger blue lasers, it just feels like a 'strong sci fi laser' color. Like the pulse rifle in base ss13, it looks similar to the disabler but you don't see anyone complaining. The base sniper reticle being red is enough for people to realize this is a more dangerous version anyways.

> Stepping through laser beams will check for eye protection, if it isn't sufficient, and the small probability chance passes, the target will scream in pain! Completely flavor. If they have enough eye protection, it will bounce off their headgear, and if said protection is BiMex patented personal shades it will create a cool visual effect.

Cool flavor. Also lore-accurate!

# Testing Photographs and Procedure

<details>
<summary>Screenshots & Videos</summary>

Put screenshots and videos here with an empty line between the screenshots and the `<details>` tags.

</details>

https://streamable.com/kyprhl

# Changelog

:cl:
add: Aiming a shot on a sniper rifle now adds a red laser beam that connects the target and the aimer. Spotting designators will emanate a faint yellow one. Both will slowly increase in alpha from 0 to max as the tracking progresses.
bal: The sniper's laser beam can be disabled with a button, but keeping it enabled reduces aiming shot time. Cloaked spotters will not cause a beam to appear.
imageadd: Completely resprited the indicators for spotting (it's now yellow) and aiming (it's now red, usually)
imageadd: The XM42B and PMC sniper rifles have more intense blue lasers and reticles.
add: Stepping through laser beams will check for eye protection, if it isn't sufficient, and the small probability chance passes, the target will scream in pain! Completely flavor. If they have enough eye protection, it will bounce off their headgear, and if said protection is BiMex patented personal shades it will create a cool visual effect.
refactor: Refactored eye_protection to have three levels, flavor, flashes, and welding. Sunglasses have flavor protection, sechuds have flash protection, welding gear has welding protection.
imageadd: The spotter's designator now has a white band on it to distinguish it from the rest.
/:cl:

### CODE REVIEW

**Comment by stanalbatross** [comment]
Architect comments:
Laser needs to inherit the sniper's alpha if it's mandatory on
Consider making it optional to decrease aimed shot timers maybe?
The eye protection thing is dumb. As fuck. Punishes players for just moving through a red line that they have no reason to believe is harmful.

**Comment by MusselMain** [comment]
Won't this Invalidate Ghillie Suits?as the xenos will know where the sniper is located? also, the laser will act like a budget Friendly fire,because it will blind anyone who sees it,cool idea however,I Really think Snipers aren't strong enough to warrant them being nerfed into Demolitionist Level Downsides,(Demolitionist have high damage but need to be close range to be effective,whereas snipers aren't as outright lethal and therefore have ghillie suits), before the spotters were added, playing sniper was very difficult on any maps with chokes, as xenos would simply reach cover before you could fire.So unless they've been buffed damage wise, 150-200 damage a shot isn't that great when the same can be achieved with anburst from a M41A MK2. so it seems abit too punishing despite being an interesting concept.(don't mind the bad spacing my laptop keyboard is messed up specifically the spacekey)

**Review by kugamo — APPROVED** [review]
Sprites will work! APPROVED.

File: code/modules/projectiles/guns/specialist.dm [hunk 1]
```diff
@@ -65,51 +78,85 @@
...
+	var/aiming_buffs
+
+	if(sniper_rifle.enable_aimed_shot_laser)
+		aiming_buffs++
+
+	if(HAS_TRAIT(target, TRAIT_SPOTTER_LAZED))
+		aiming_buffs++
+
+	if(aiming_buffs == 1)
+		f_aiming_time *= 0.6
```
Comment (Diegoflores31) [inline code/modules/projectiles/guns/specialist.dm:112]: Undocumented nerf to spotter . It used to be 0.5 + wicked way to apply the stuff lol

Comment (carlarctg) [reply code/modules/projectiles/guns/specialist.dm:112]: stan knows about it

figure out a better way loser!

Comment (carlarctg) [reply code/modules/projectiles/guns/specialist.dm]: image

ok i did that instead

**Comment by harryob** [comment]
this bot stupid asf

File: code/datums/beam.dm [hunk 2]
```diff
@@ -143,6 +143,44 @@
...
+	var/rand_laser_power = rand(EYE_PROTECTION_FLAVOR, strength)
+	if(rand_laser_power > laser_protection)
+		//ouch!
+		INVOKE_ASYNC(moving_human, /mob/proc/emote, "pain")
+		visible_message(SPAN_DANGER("[moving_human] screams out in pain as \the [src] moves across their eyes!"), SPAN_NOTICE("Aurgh!!! \The [src] moves across your unprotected eyes for a split-second!"))
+	else
+		if(HAS_TRAIT(moving_human, TRAIT_BIMEX))
+			visible_message(SPAN_NOTICE("[moving_human]'s BiMex© personal shades shine for a moment as \the [src] passes over them."), SPAN_NOTICE("Your BiMex© personal shades shine for a moment as \the [src] passes over them."))
+			//drip = bonus balloonchat
+			moving_human.balloon_alert_to_viewers(SPAN_NOTICE("[moving_human]'s BiMex© personal shades shine for a moment as \the [src] passes over them."), SPAN_NOTICE("Your BiMex© personal shades reflect \the [src]."))
```
Comment (stanalbatross) [inline code/datums/beam.dm:168]: this first message is way too long and span classes don't work in balloon chat iirc

File: code/modules/projectiles/guns/specialist.dm [hunk 3]
```diff
@@ -13,9 +13,23 @@
...
+	var/sniper_beam_icon = "laser_beam"
+
+/obj/item/weapon/gun/rifle/sniper/get_examine_text(mob/user)
+	. = ..()
+	if(!has_aimed_shot)
+		return
+	var/dat = ""
+	dat += SPAN_NOTICE("This weapon has an unique ability, Aimed Shot, allowing it to deal great damage after a windup. ")
+	dat += SPAN_BOLDNOTICE("Additionally, the aimed shot can be sped up with a tracking laser, which is enabled by default but may be disabled.")
+	. += dat
```
Comment (stanalbatross) [inline code/modules/projectiles/guns/specialist.dm:28]: genuinely schizo code, make it 1 line normally

File: code/modules/projectiles/guns/specialist.dm [hunk 4]
```diff
@@ -167,6 +218,54 @@
...
+
+	button.overlays.Cut()
+	var/image/IMG = image(icon, button, icon_state)
+	button.overlays += IMG
+
+/datum/action/item_action/specialist/toggle_laser/can_use_action()
+	var/mob/living/carbon/human/human = owner
+	var/obj/item/weapon/gun/rifle/sniper/sniper_rifle = holder_item
+
+	if(!istype(human) || human.is_mob_incapacitated())
```
Comment (stanalbatross) [inline code/modules/projectiles/guns/specialist.dm:248]: istype human :/

File: code/modules/projectiles/guns/specialist.dm [hunk 5]
```diff
@@ -167,6 +218,54 @@
...
+		return FALSE
+
+	if(sniper_rifle != human.r_hand && sniper_rifle != human.l_hand)
```
Comment (stanalbatross) [inline code/modules/projectiles/guns/specialist.dm:251]: there is a proper proc for this

File: code/modules/projectiles/guns/specialist.dm [hunk 6]
```diff
@@ -167,6 +218,54 @@
...
+		to_chat(human, SPAN_WARNING("How do you expect to do this without your sniper rifle?"))
+		return FALSE
+	return TRUE
+
+/datum/action/item_action/specialist/toggle_laser/action_activate()
+	var/mob/living/carbon/human/human = owner
+	var/obj/item/weapon/gun/rifle/sniper/sniper_rifle = holder_item
+
+	if(sniper_rifle != human.r_hand && sniper_rifle != human.l_hand)
```
Comment (stanalbatross) [inline code/modules/projectiles/guns/specialist.dm:260]: same

File: code/modules/projectiles/guns/specialist.dm [hunk 7]
```diff
@@ -167,6 +218,54 @@
...
+
+/datum/action/item_action/specialist/toggle_laser/can_use_action()
+	var/mob/living/carbon/human/human = owner
+	var/obj/item/weapon/gun/rifle/sniper/sniper_rifle = holder_item
+
+	if(!istype(human) || human.is_mob_incapacitated())
+		return FALSE
+
+	if(sniper_rifle != human.r_hand && sniper_rifle != human.l_hand)
+		to_chat(human, SPAN_WARNING("How do you expect to do this without your sniper rifle?"))
```
Comment (stanalbatross) [inline code/modules/projectiles/guns/specialist.dm:252]: must hold it in hand, clarify this message

**Review by stanalbatross — CHANGES_REQUESTED** [review]

File: code/modules/projectiles/guns/specialist.dm (hunk 3 above)
Comment (carlarctg) [reply code/modules/projectiles/guns/specialist.dm:28]: bro i copied it from YOUR lever action code

File: code/modules/projectiles/guns/specialist.dm (hunk 4 above)
Comment (carlarctg) [reply code/modules/projectiles/guns/specialist.dm:248]: dont see u fixin thes typecheks on tg ports

File: code/modules/projectiles/guns/specialist.dm (hunk 5 above)
Comment (carlarctg) [reply code/modules/projectiles/guns/specialist.dm:251]: and you don't say which one it is okay man. okay .

File: code/modules/projectiles/guns/specialist.dm (hunk 7 above)
Comment (carlarctg) [reply code/modules/projectiles/guns/specialist.dm:252]: done

File: code/modules/projectiles/guns/specialist.dm (hunk 6 above)
Comment (carlarctg) [reply code/modules/projectiles/guns/specialist.dm:260]: done

File: code/modules/projectiles/guns/specialist.dm (hunk 5 above)
Comment (carlarctg) [reply code/modules/projectiles/guns/specialist.dm:251]: dobne

File: code/modules/projectiles/guns/specialist.dm (hunk 4 above)
Comment (carlarctg) [reply code/modules/projectiles/guns/specialist.dm:248]: donbe

File: code/modules/projectiles/guns/specialist.dm (hunk 3 above)
Comment (carlarctg) [reply code/modules/projectiles/guns/specialist.dm:28]: yeaaaaa

File: code/datums/beam.dm (hunk 2 above)
Comment (carlarctg) [reply code/datums/beam.dm:168]: okie dokie

Comment (carlarctg) [reply code/datums/beam.dm]: fixed
image

File: code/datums/beam.dm [hunk 8]
```diff
@@ -143,6 +143,44 @@
...
+/obj/effect/ebeam/laser/intense
+	name = "intense laser beam"
+	alpha = 255
+	strength = EYE_PROTECTION_FLASH
+	probability = 35
+
+/obj/effect/ebeam/laser/weak
+	name = "weak laser beam"
+	alpha = 150
+	strength = 1
```
Comment (harryob) [inline code/datums/beam.dm:181]:
```suggestion
	strength = EYE_PROTECTION_FLAVOR
```

**Review by harryob — CHANGES_REQUESTED** [review]

**Pull request merged**